util/lmdb/src/metadata_store.rs

```diff
@@ -0,0 +1,216 @@
+    VersionIncompatible(u64, u64),
+
+    /// Serialization
+    Serialization,
+
+    /// Deserialization
+    Deserialization,
+}
+
+impl From<LmdbError> for MetadataStoreError {
+    fn from(src: LmdbError) -> Self {
+        Self::Lmdb(src)
+    }
+}
+
+impl From<mc_util_serial::DecodeError> for MetadataStoreError {
+    fn from(_: mc_util_serial::DecodeError) -> Self {
+        Self::Deserialization
+    }
+}
+
+impl From<mc_util_serial::EncodeError> for MetadataStoreError {
+    fn from(_: mc_util_serial::EncodeError) -> Self {
+        Self::Serialization
+    }
+}
+
+/// A trait that defines the per-db settings for a MetadataStore.
+pub trait MetadataStoreSettings: Default {
+    /// Default database version. This should be bumped when breaking changes are introduced.
+    /// If this is properly maintained, we could check during database opening for any
+    /// incompatibilities, and either refuse to open or perform a migration.
+    const LATEST_VERSION: u64;
+
+    /// The current crate version that manages the database.
+    const CRATE_VERSION: &'static str;
+
+    /// LMDB Database name to use for storing the metadata information.
+    const DB_NAME: &'static str;
+
+    /// Check if a given version is compatible with the latest version.
+    /// The default implementation assumes only the latest version is compatible.
+    fn is_compatible_with_latest(
+        metadata_version: &MetadataVersion<Self>,
+    ) -> Result<(), MetadataStoreError> {
+        let latest = MetadataVersion::<Self>::latest();
+        if metadata_version.database_format_version != latest.database_format_version {
+            Err(MetadataStoreError::VersionIncompatible(
+                metadata_version.database_format_version,
+                latest.database_format_version,
+            ))
+        } else {
+            Ok(())
+        }
+    }
+}
+
+/// Metadata information stored inside the LMDB database.
+#[derive(Clone, Message)]
+struct StoredMetadataVersion {
+    /// Database format version.
+    #[prost(uint64)]
+    pub database_format_version: u64,
+
+    /// Crate version that created the database. This could be bumped by a migration in a future
+    /// release.
+    #[prost(string)]
+    pub created_by_crate_version: String,
+}
+impl<S: MetadataStoreSettings> From<&MetadataVersion<S>> for StoredMetadataVersion {
+    fn from(src: &MetadataVersion<S>) -> Self {
+        Self {
+            database_format_version: src.database_format_version,
+            created_by_crate_version: src.created_by_crate_version.clone(),
+        }
+    }
+}
+
+/// Metadata information about a database instance.
+#[derive(Clone, Debug)]
+pub struct MetadataVersion<S: MetadataStoreSettings> {
+    /// Database format version.
+    pub database_format_version: u64,
+
+    /// Crate version that created the database. This could be bumped by a migration in a future
+    /// release.
+    pub created_by_crate_version: String,
+
+    _s: S,
```

Comment:
Ideally this would've been `PhantomData<S>` instead, but Rust doesn't like that (https://github.com/rust-lang/rust/issues/46969 might be related).